File: loja-cpp-catch2/src/carrinho.cpp
#include "carrinho.hpp"

void Carrinho::adicionar(const Produto& produto) {
    produtos_.push_back(produto);
}

double Carrinho::total() const {
    double soma = 0.0;

    for (const auto& p : produtos_) {
        soma += p.preco();
    }

    return soma;
}

std::size_t Carrinho::quantidade() const {
    return produtos_.size();
}

File: loja-typescript/src/carrinho.ts
import { Produto } from "./produto";

export class Carrinho {
  private produtos: Produto[] = [];

  adicionar(produto: Produto): void {
    this.produtos.push(produto);
  }

  total(): number {
    return this.produtos.reduce((soma, p) => soma + p.preco, 0);
  }

  quantidade(): number {
    return this.produtos.length;
  }
}
